src/toBeSorted/d_emitter.cpp

```diff
@@ -0,0 +1,1298 @@
+    mFlags = 0;
+    setImmortal();
+    dJEffManager_c::sPlayingEffectsList.append(this);
+    onFlag(EMITTER_0x1);
+    if (bFlags) {
+        onFlag(EMITTER_0x4);
+    }
+    if (dParticle::mgr_c::GetInstance()->getResUserWork(resourceId) & 0x4000) {
+        onFlag(EMITTER_0x10);
+    }
+}
+
+bool EffectsStruct::areAllEmittersDone() {
+    bool allDone = true;
+
+    if (mpEmitterHead != nullptr) {
+        if (checkFlag(EMITTER_Fading) && mFadeTimer != 0) {
+            return false;
+        }
+        for (JPABaseEmitter *emitter = mpEmitterHead; emitter != nullptr; emitter = GetNextEmitter(emitter)) {
+            bool canDelete = emitter->isEnableDeleteEmitter();
+            if (!canDelete) {
+                if (emitter->mMaxFrame == 0) {
+                    emitter->becomeInvalidEmitter();
+                }
+                allDone = false;
+            }
+        }
+    }
+
+    return allDone;
+}
+
+void EffectsStruct::execute() {
+    if (mpOwner != nullptr && (mpOwner->delete_request || mpOwner->lifecycle_state == fBase_c::TO_BE_DELETED)) {
+        mpOwner = nullptr;
+    }
+    if (dJEffManager_c::shouldBePaused(mpOwner)) {
+        if (!checkFlag(EMITTER_0x10)) {
+            stopCalcEmitters();
+            if (mpOwner != nullptr && !mpOwner->isBasePropertyFlag(0x100)) {
+                if (!mpEmitterHead->checkStatus(JPAEmtrStts_StopDraw)) {
+                    onFlag(EMITTER_0x20);
+                    stopDrawParticles();
+                }
+            } else {
+                if (checkFlag(EMITTER_0x20)) {
+                    offFlag(EMITTER_0x20);
+                    playDrawParticles();
+                }
+            }
+        }
+        onFlag(EMITTER_0x1);
+    } else {
+        playCalcEmitters();
+        if (checkFlag(EMITTER_0x20)) {
+            offFlag(EMITTER_0x20);
+            playDrawParticles();
+        }
+        if (checkFlag(EMITTER_Fading) && mFadeTimer != 0) {
+            mFadeTimer--;
+        }
+    }
+}
+
+bool EffectsStruct::getOwnerPolyAttrs(s32 *pOut1, s32 *pOut2) {
+    if (mpOwner != nullptr && mpOwner->group_type == fBase_c::ACTOR) {
+        dAcBase_c *actor = static_cast<dAcBase_c *>(mpOwner);
+        *pOut1 = actor->polyAttr0;
+        *pOut2 = actor->polyAttr1;
+        return true;
+    } else {
+        *pOut1 = 0;
+        *pOut2 = 0;
+        return false;
+    }
+}
+
+void EffectsStruct::realizeAlpha() {
+    setGlobalAlpha(mFadeTimer * (255.0f / mFadeDuration));
+}
+
+void EffectsStruct::setFading(u8 lifetime) {
+    if (!checkFlag(EMITTER_Fading)) {
+        mFadeTimer = lifetime;
+        onFlag(EMITTER_Fading);
+        mFadeDuration = lifetime;
+    }
+}
+
+bool dParticleFogProc_c::create(u32 idx, s32 prioOpa, s32 prioXlu, mHeapAllocator_c *alloc) {
+    if (!d3d::UnkProc::create(prioOpa, prioXlu, alloc)) {
+        return false;
+    }
+    mIdx = idx;
+    // idxes 0, 2, 4, 6, 10
+    field_0x1C = idx <= 10 && ((1 << idx) & 0b10001010101) != 0;
```

Comment:
Does this need to be represented as such?
I know the usual trick is to do `idx == 0 || idx == 2 || idx == 4 || idx == 6 || idx == 10`

Reply:
I've never seen that before, cool! I only knew about the `(unsigned)(X - Y) < Z` trick for ranges, but not about these bit tricks.